src/paradigma/heart_rate/tfd.py

```diff
@@ -0,0 +1,553 @@
+    """
+    Generate the Doppler-lag kernel based on kernel type and parameters.
+
+    Parameters:
+    -----------
+    kern_type : str
+        Type of kernel (e.g., 'wvd', 'swvd', 'pwvd', etc.).
+    kern_params : dict
+        Parameters for the kernel.
+    N : int
+        Signal length.
+    lag_index : int
+        Current lag index.
+
+    Returns:
+    --------
+    g : ndarray
+        Doppler-lag kernel for the given lag.
+    """
+    g = np.zeros(N, dtype=complex)  # Initialize the kernel
+
+    # Get kernel based on the type
+    g = get_kern(g, lag_index, kern_type, kern_params, N)
+
+    return np.real(g) # All kernels are real valued
+
+
+def get_kern(g, lag_index, kern_type, kern_params, N):
+    """
+    Get the kernel based on the provided kernel type.
+
+    Parameters:
+    -----------
+    g : ndarray
+        Kernel to be filled.
+    lag_index : int
+        Lag index.
+    kern_type : str
+        Type of kernel ('wvd', 'swvd', 'pwvd', 'sep').
+    kern_params : dict
+        Parameters for the kernel.
+    N : int
+        Signal length.
+
+    Returns:
+    --------
+    g : ndarray
+        Kernel function at the current lag.
+    """
+    l = len(kern_params)
+
+    if kern_type == 'wvd':
+        g[:] = 1 # WVD kernel is the equal to 1 for all lags
+
+    elif kern_type == 'swvd':
+        key = 'lag'  # Key for the lag-independent kernel
+        # Smoothed Wigner-Ville Distribution (Lag Independent kernel)
+        if l < 2:
+            raise ValueError("Need at least two window parameters for SWVD")
+        win_length = kern_params['win_length']
+        win_type = kern_params['win_type']
+        win_param = kern_params['win_param'] if l >= 3 else 0
+        win_param2 = kern_params['win_param2'] if l >= 4 else 1
+
+        G1 = get_window(win_length, win_type, win_param)
+        G1 = pad_window(G1, N)
+
+        # Define window in the time domain or Doppler domain
+        if win_param2 == 0:
+            G1 = np.fft.fft(G1)
+            G1 /= G1[0]
+
+        g[:] = G1 # Assign the window to the kernel
+
+    elif kern_type == 'pwvd':
+        key = 'doppler'
+        # Pseudo-Wigner-Ville Distribution (Doppler Independent kernel)
+        if l < 2:
+            raise ValueError("Need at least two window parameters for PWVD") 
+        win_length = kern_params['win_length']
+        win_type = kern_params['win_type']
+        win_param = kern_params['win_param'] if l >= 3 else 0
+        win_param2 = kern_params['win_param2'] if l >= 4 else 0
+
+        G2 = get_window(win_length, win_type, win_param) # Generate the window, same per lag iteration
+        G2 = pad_window(G2, N)      # Zero-pad the window to the length of the signal
+        G2 = G2[lag_index]          # Extract the lag_index-th element of the window --> can this be done more efficiently? Since we only need one element of the window and the padding is similar for every iteration
+
+        g[:] = G2 # Assign the window to the kernel
```

Comment:
I think you could combine these:

First:
```
if kern_type not in [X, Y, Z, A, B, C]:
    raise("You should pick one of ...")
```

Then
```
if kern_type == 'wvd':
    ...
```

```
elif kern_type == 'sep': 
    ....
```

```
else:
    if l < 2:
        raise ValueError("Need at least two window parameters for {kern_type.upper()}")

    if kern_type == 'swvd':
        key = 'lag'
    else:
        key = 'doppler'
        
      win_length = kern_params['win_length']
      win_type = kern_params['win_type']
      win_param = kern_params['win_param'] if l >= 3 else 0
      win_param2 = kern_params['win_param2'] if l >= 4 else 1

      G = get_window(win_length, win_type, win_param)
      G = pad_window(G, N)

      if kern_type == 'sqvd' and win_param2 == 0:
          G = np.fft.fft(G)
          G /= G[0]
    elif kern_type == 'pwvd':
        G = G[lag_index]    

      g[:] = G 
```

And I don't think you have to assign `key` in this loop because you don't seem to be using it.

Reply:
Agree, i did rewrite the function accordingly